File: sirma-platform/ui2/src/administration/model-management/model/model-control.js
import {ModelBase} from 'administration/model-management/model/model-base';
import {ModelControlParam} from 'administration/model-management/model/model-control-param';
import {ModelList} from 'administration/model-management/model/model-list';

/**
 * Represents a model of a control extending the {@link ModelBase} class.
 * A model controls are bound to {@link ModelField}.
 *
 * @author svelikov
 */
export class ModelControl extends ModelBase {

  constructor(id) {
    super(id);
    this.controlParams = new ModelList();
  }

  getControlParams() {
    return this.controlParams.getModels();
  }

  getOwnControlParams() {
    return this.getOwnModels(this.getControlParams());
  }

  getDirtyControlParams() {
    return ModelBase.getDirtyModels(this.getOwnControlParams());
  }

  getControlParam(id) {
    return this.controlParams.getModel(id);
  }

  addControlParam(controlParam) {
    if (controlParam instanceof ModelControlParam) {
      this.controlParams.insert(controlParam);
    }
    return this;
  }

  removeControlParam(controlParam) {
    let id = controlParam.getId();
    if (this.getControlParam(id)) {
      this.controlParams.remove(id);
    }
    return this;
  }

  isValid() {
    return super.isValid() && ModelBase.areModelsValid(this.getDirtyControlParams());
  }

  isDirty() {
    return super.isDirty() || ModelBase.areModelsDirty(this.getOwnControlParams());
  }

  copyFrom(src) {
    super.copyFrom(src);
    this.controlParams.copyFrom(src.controlParams);
    return this;
  }

  isOwningModels() {
    return super.isOwningModels() || !!this.getOwnControlParams().length;
  }
}

File: sirma-platform/platform/seip-parent/model-management/model-management-api/src/main/java/com/sirma/sep/model/management/ModelControl.java
package com.sirma.sep.model.management;

import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Optional;
import java.util.function.Function;

import com.sirma.itt.seip.Copyable;
import com.sirma.sep.model.management.definition.DefinitionModelAttributes;
import com.sirma.sep.model.management.meta.ModelMetaInfo;

/**
 * Holds definition information of controls. The controls are holders of control parameters or
 * {@link ModelControlParam}s.
 *
 * @author Stella Djulgerova.
 */
public class ModelControl extends AbstractModelNode<ModelControl, ModelField> implements Copyable<ModelControl> {

	static final String MODEL_TYPE = "control";

	private Map<String, ModelControlParam> controlParams;

	@Override
	protected String getLabelAttributeKey() {
		return DefinitionModelAttributes.LABEL;
	}

	@Override
	public ModelControl createCopy() {
		ModelControl copyOfModelControl = copyNodeTo(new ModelControl());
		getControlParams().stream().map(ModelControlParam::createCopy).forEach(copyOfModelControl::addControlParam);
		return copyOfModelControl;
	}

	@Override
	public int hashCode() {
		return super.hashCode() * 27;
	}

	@Override
	protected String getTypeName() {
		return MODEL_TYPE;
	}

	@Override
	public boolean equals(Object o) {
		return super.equals(o);
	}

	@Override
	protected Function<String, Optional<ModelControl>> getRemoveFunction() {
		return getContext()::removeControl;
	}

	@Override
	public Object walk(Path step) {
		if (ModelControlParam.MODEL_TYPE.equals(step.getName())) {
			return step.proceed(getOrCreateControlParam(step.getValue()));
		}
		return super.walk(step);
	}

	@Override
	public Map<String, ModelMetaInfo> getAttributesMetaInfo() {
		return getModelsMetaInfo().getControlsMapping();
	}

	@Override
	public boolean isEmpty() {
		return true;
	}

	public void setControlParams(Map<String, ModelControlParam> controlParams) {
		this.controlParams = controlParams;
	}

	private ModelControlParam getOrCreateControlParam(String id) {
		return getControlParamsMap().computeIfAbsent(id, this::createControlParam);
	}

	private ModelControlParam createControlParam(String id) {
		ModelControlParam controlParamModel = new ModelControlParam();
		controlParamModel.setId(id);
		controlParamModel.setModelsMetaInfo(getModelsMetaInfo());
		controlParamModel.setContext(this);
		controlParamModel.setDetachedModelNodesStore(getDetachedModelNodesStore());
		return controlParamModel;
	}

	public Collection<ModelControlParam> getControlParams() {
		return getControlParamsMap().values();
	}

	public Map<String, ModelControlParam> getControlParamsMap() {
		if (controlParams == null) {
			controlParams = new LinkedHashMap<>();
		}
		return controlParams;
	}

	public ModelControlParam addControlParam(ModelControlParam controlParam) {
		controlParam.setContext(this);
		controlParam.setDetachedModelNodesStore(getDetachedModelNodesStore());
		getDetachedModelNodesStore().removeDetached(controlParam);
		return getControlParamsMap().put(controlParam.getId(), controlParam);
	}

	public Optional<ModelControlParam> removeControlParam(String controlParamId) {
		return Optional.ofNullable(this.getControlParamsMap().remove(controlParamId));
	}
}
